File: server/webapp/WEB-INF/rails.new/app/controllers/api_v2/base_controller.rb
##########################GO-LICENSE-START################################
# Copyright 2015 ThoughtWorks, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##########################GO-LICENSE-END##################################

module ApiV2
  class BaseController < ::ApplicationController

    class BadRequest < StandardError
    end

    include AuthenticationHelper

    FORMATS                = [:json_hal_v2]
    DEFAULT_FORMAT         = FORMATS.last
    DEFAULT_ACCEPTS_HEADER = Mime[DEFAULT_FORMAT].to_s

    skip_before_filter :verify_authenticity_token
    before_filter :verify_content_type_on_post
    before_filter :set_default_response_format

    def set_default_response_format
      request.format = DEFAULT_FORMAT unless params[:format]
    end

    def redirect_json_hal(url, status=302)
      json = {
        _links: {
          redirect: {
            href: url
          }
        }
      }

      response.status = status
      render json_hal_v2: json, location: url
    end

    rescue_from RecordNotFound, with: :render_not_found_error
    rescue_from BadRequest,     with: :render_bad_request

    class << self
      def default_accepts_header
        DEFAULT_ACCEPTS_HEADER
      end
    end

    protected

    def to_tristate(var)
      TriState.from(var.to_s)
    rescue => e
      raise BadRequest.new(e.message)
    end

    def render_http_operation_result(result)
      if result.instance_of?(HttpOperationResult)
        render json_hal_v2: { message: result.detailedMessage().strip }, status: result.httpCode()
      else
        render json_hal_v2: { message: result.message(Spring.bean('localizer')).strip }, status: result.httpCode()
      end
    end
  end
end

File: api/api/src/main/java/com/thoughtworks/go/server/api/BaseController.java
/*
 * Copyright 2018 ThoughtWorks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.thoughtworks.go.server.api;

import org.apache.commons.lang.StringUtils;
import org.apache.http.client.utils.URLEncodedUtils;
import org.apache.http.message.BasicNameValuePair;
import org.springframework.http.HttpStatus;
import spark.Request;
import spark.Response;

import javax.activation.MimeType;
import javax.activation.MimeTypeParseException;
import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

import static org.apache.commons.lang.StringUtils.isBlank;
import static spark.Spark.halt;

public abstract class BaseController implements ControllerMethods {
    private static final Set<String> UPDATE_HTTP_METHODS = new HashSet<>(Arrays.asList("PUT", "POST", "PATCH"));

    protected final ApiVersion apiVersion;
    protected final String mimeType;

    protected BaseController(ApiVersion apiVersion) {
        this.apiVersion = apiVersion;
        this.mimeType = apiVersion.mimeType();
    }

    public String controllerPath(String... paths) {
        if (paths == null || paths.length == 0) {
            return controllerBasePath();
        } else {
            return (controllerBasePath() + "/" + StringUtils.join(paths, '/')).replaceAll("//", "/");
        }
    }

    public String controllerPath(Map<String, String> params) {
        if (params == null || params.isEmpty()) {
            return controllerBasePath();
        } else {
            List<BasicNameValuePair> queryParams = params.entrySet().stream().map(new Function<Map.Entry<String, String>, BasicNameValuePair>() {
                @Override
                public BasicNameValuePair apply(Map.Entry<String, String> entry) {
                    return new BasicNameValuePair(entry.getKey(), entry.getValue());
                }
            }).collect(Collectors.toList());
            return controllerBasePath() + '?' + URLEncodedUtils.format(queryParams, "utf-8");
        }
    }

    protected abstract String controllerBasePath();

    public abstract void setupRoutes();

    protected void setContentType(Request req, Response res) {
        res.raw().setCharacterEncoding("utf-8");
        res.header("Content-Type", mimeType);
    }

    protected String messageJson(Exception ex) {
        return MessageJson.create(ex.getMessage());
    }

    protected void verifyContentType(Request request, Response response) {
        if (!UPDATE_HTTP_METHODS.contains(request.requestMethod().toUpperCase())) {
            return;
        }

        if (request.contentLength() >= 1 && !isJsonContentType(request)) {
            throw halt(HttpStatus.UNSUPPORTED_MEDIA_TYPE.value(), MessageJson.create("You must specify a 'Content-Type' of 'application/json'"));
        }

        if ("chunked".equalsIgnoreCase(request.headers("Transfer-Encoding")) && !isJsonContentType(request)) {
            throw halt(HttpStatus.UNSUPPORTED_MEDIA_TYPE.value(), MessageJson.create("You must specify a 'Content-Type' of 'application/json'"));
        }
    }

    protected boolean isJsonContentType(Request request) {
        String mime = request.headers("Content-Type");
        if (isBlank(mime)) {
            return false;
        }
        try {
            return new MimeType(mime).getBaseType().equals("application/json");
        } catch (MimeTypeParseException e) {
            return false;
        }
    }

    public String getMimeType() {
        return mimeType;
    }
}
